File: pis2e.js
(function (ext) {
    var socket = null;
    var connected = false;
    var myStatus = 1;
    var myMsg = 'not_ready';

    ext.connect = function (callback) {
        window.socket = new WebSocket("ws://127.0.0.1:9000");
        window.socket.onopen = function () {
            var msg = JSON.stringify({
                "command": "ready"
            });

            window.socket.send(msg);
            myStatus = 2;
            myMsg = 'ready';
            connected = true;

            window.setTimeout(function() {
                callback();
            }, 1000);

        };

        window.socket.onmessage = function (message) {
        };

        window.socket.onclose = function (e) {
            console.log("Connection closed.");
            socket = null;
            connected = false;
            myStatus = 1;
            myMsg = 'not_ready'
        };
    };

    ext._shutdown = function () {
        var msg = JSON.stringify({
            "command": "shutdown"
        });
        window.socket.send(msg);
    };

    ext._getStatus = function (status, msg) {
        return {status: myStatus, msg: myMsg};
    };

    ext.sendmsg = function (msg) {
        if (connected == false) {
            alert("Server Not Connected");
        }
        // validate the pin number for the mode
        var msg = JSON.stringify({
            "command": 'sendmsg', 'msg': msg
        });
        console.log(msg);
        window.socket.send(msg);
    };

    var descriptor = {
        blocks: [
            ["w", 'Connect to WebSocket Server', 'connect'],
            [" ", "Send a message: %s", "sendmsg", "Hello World!"]
        ],
        "menus": {
            "high_low": ["0", "1"]
        },
        url: 'https://github.com/motojin/pis2e-boilerplate'
    };

    ScratchExtensions.register('pis2e by motojin', descriptor, ext);
})({});

File: pis2e.py
import json
import sys
from SimpleWebSocketServer import SimpleWebSocketServer, WebSocket


class MarudaClub(WebSocket):
   
    def handleMessage(self):
        payload = json.loads(self.data)
        print(payload)
        client_cmd = payload['command']
        if client_cmd == 'sendmsg':
            msg = payload['msg']
            print(msg)
        elif client_cmd == 'ready':
            pass
        else:
            print("Unknown command received", client_cmd)

    def handleConnected(self):
        print(self.address, 'connected')

    def handleClose(self):
        print(self.address, 'closed')


def run_server():
    server = SimpleWebSocketServer('', 9000, MarudaClub)
    server.serveforever()


if __name__ == "__main__":
    try:
        run_server()
    except KeyboardInterrupt:
        sys.exit(0)

